heapster-saw/src/Verifier/SAW/Heapster/Implication.hs

```diff
@@ -4584,19 +4620,48 @@ proveEqH psubst e mb_e = case (e, mbMatch mb_e) of
   (PExpr_LLVMWord e', [nuMP| PExpr_LLVMWord mb_e' |]) ->
     fmap PExpr_LLVMWord <$> proveEqH psubst e' mb_e'
 
-  -- Prove e = N*z + M where e - M is a multiple of N by setting z = (e-M)/N
-  (_, [nuMP| PExpr_BV [BVFactor (BV.BV mb_n) z] (BV.BV mb_m) |])
-    | Left memb <- mbNameBoundP z
-    , Nothing <- psubstLookup psubst memb
-    , bvIsZero (bvMod (bvSub e (bvInt $ mbLift mb_m)) (mbLift mb_n)) ->
-      setVarM memb (bvDiv (bvSub e (bvInt $ mbLift mb_m)) (mbLift mb_n)) >>>
-      pure (SomeEqProofRefl e)
+  -- Prove e = L_1*y_1 + ... + L_k*y_k + N*z + M where z is an unset variable,
+  -- each y_i is either a set variable with value e_i or an unbound variable
+  -- with e_i = y_i, and e - (L_1*e_1 + ... + L_k*e_k + M) is divisible by N,
+  -- by setting z = (e - (L_1*e_1 + ... + L_k*e_k + M))/N
+  (_, [nuMP| PExpr_BV mb_factors (BV.BV mb_m) |])
+    | Just (n, memb, e_factors) <- getUnsetBVFactor mb_factors
+    , e' <- bvSub e (bvAdd e_factors (bvInt $ mbLift mb_m))
+    , bvIsZero (bvMod e' n) ->
+      setVarM memb (bvDiv e' n) >>> pure (SomeEqProofRefl e)
 
   -- FIXME: add cases to prove struct(es1)=struct(es2)
 
   -- Otherwise give up!
   _ -> proveEqFail e mb_e
 
+  where -- If there is exactly one 'BVFactor' in a list of 'BVFactor's which is
```

Comment:
Please put these two helper functions into Permissions.hs and make them top-level functions